ttnn/cpp/ttnn/operations/conv/conv2d/conv2d_utils.cpp

```diff
@@ -782,39 +853,75 @@ ttnn::operations::matmul::MatmulProgramConfig determine_matmul_op_config_from_co
+    uint32_t out_subblock_w_ntiles, uint32_t out_subblock_h_ntiles,
+    uint32_t max_num_subblock, uint32_t act_block_h_ntiles, uint32_t act_block_h_ntiles_padded,
+    uint32_t weight_block_w_ntiles, uint32_t out_block_h_ntiles, uint32_t max_subblock_h)
+{
+    if ((out_subblock_w_ntiles * out_subblock_h_ntiles <= max_num_subblock / 2) and
+        (out_subblock_w_ntiles == weight_block_w_ntiles) and (act_block_h_ntiles == out_block_h_ntiles)) {
+        uint32_t num_subblock_h = act_block_h_ntiles / out_subblock_h_ntiles;
+        uint32_t num_iter = max_subblock_h - out_subblock_h_ntiles;
+        uint32_t new_out_subblock_h = out_subblock_h_ntiles;
+        uint32_t preferred_out_subblock_h = out_subblock_h_ntiles;
+
+        for (uint32_t i = 0; i < num_iter; ++i) {
+            new_out_subblock_h += 1;
+            uint32_t new_num_subblock_h = (act_block_h_ntiles + new_out_subblock_h - 1) / new_out_subblock_h;
+
+            if (new_num_subblock_h < num_subblock_h and
+                (out_subblock_w_ntiles * new_out_subblock_h <= max_num_subblock)) {
+                num_subblock_h = new_num_subblock_h;
+                preferred_out_subblock_h = new_out_subblock_h;
+            }
+        }
+        return {preferred_out_subblock_h, preferred_out_subblock_h * num_subblock_h};
+    }
+    return {out_subblock_h_ntiles, act_block_h_ntiles};
+}
+
+Conv2dConfig adjust_conv_op_config_for_auto_shard_if_necessary(
+    const Conv2dConfig& conv_config_,
     bool is_mm_conv,
     uint32_t batch_size,
     uint32_t in_channels,
     uint32_t out_channels,
     uint32_t output_height,
     uint32_t output_width,
     uint32_t weights_width,
+    uint32_t input_height,
     uint32_t input_width,
+    uint32_t groups,
+    std::array<uint32_t, 2> kernel_size,
     const CoreCoord& compute_grid_size,
-    Conv2dConfig& conv_config,
-    Layout input_tensor_layout,
+    const DeviceComputeKernelConfig& compute_config,
     std::optional<const MemoryConfig> input_memory_config) {
 
+    Conv2dConfig conv_config = conv_config_;
     // If the input tensor is already sharded, or the conv_config has a specified shard layout, we don't need to do anything.
     if ((input_memory_config.has_value() && input_memory_config.value().is_sharded()) || conv_config.shard_layout.has_value()) {
-        return;
+        return conv_config;
     }
 
     ShardOrientation shard_orientation =
         conv_config.transpose_shards ? ShardOrientation::COL_MAJOR : ShardOrientation::ROW_MAJOR;
     conv_config.shard_layout = select_shard_spec(
+        conv_config,
+        compute_config,
         is_mm_conv,
         batch_size,
         in_channels,
         out_channels,
         output_height,
         output_width,
         weights_width,
+        input_height,
         input_width,
-        shard_orientation,
+        groups,
+        kernel_size,
         compute_grid_size);
 
+    auto input_tensor_layout = Layout::ROW_MAJOR;
```

Comment:
why is this hardcoded now to  Layout::ROW_MAJOR?

Reply:
The output from halo, i.e, the input to the conv micro op, must always be Row Major.

Reply:
This is not output from halo!